src/linear_algebra/tensor_algebra.lean

```diff
@@ -0,0 +1,239 @@
+  smul_def' := λ _ _, rfl }
+
+/--
+The canonical linear map `M →ₗ[R] tensor_algebra R M`.
+-/
+def ι : M →ₗ[R] (tensor_algebra R M) :=
+{ to_fun := λ m, quot.mk _ m,
+  map_add' := λ x y, quot.sound rel.add_lin,
+  map_smul' := λ r x, quot.sound rel.smul_lin }
+
+/--
+Given a linear map `f : M → A` where `A` is an `R`-algebra,
+`lift R M f` is the unique lift of `f` to a morphism of `R`-algebras
+`tensor_algebra R M → A`.
+-/
+def lift {A : Type u₃} [semiring A] [algebra R A] (f : M →ₗ[R] A) : tensor_algebra R M →ₐ[R] A :=
+{ to_fun := λ a, quot.lift_on a (lift_fun _ _ f) $ λ a b h,
+  begin
+    induction h,
+    { change f _ = f _ + f _, simp, },
+    { change f _ = (algebra_map _ _ _) * f _, rw linear_map.map_smul,
+      exact algebra.smul_def h_r (f h_a) },
+    { exact (algebra_map R A).map_add h_r h_s, },
+    { exact (algebra_map R A).map_mul h_r h_s },
+    { apply algebra.commutes },
+    { change _ + _ + _ = _ + (_ + _), rw add_assoc },
+    { change _ + _ = _ + _, rw add_comm, },
+    { change (algebra_map _ _ _) + lift_fun R M f _ = lift_fun R M f _, simp, },
+    { change _ * _ * _ = _ * (_ * _), rw mul_assoc },
+    { change (algebra_map _ _ _) * lift_fun R M f _ = lift_fun R M f _, simp, },
+    { change lift_fun R M f _ * (algebra_map _ _ _)= lift_fun R M f _, simp, },
+    { change _ * (_ + _) = _ * _ + _ * _, rw left_distrib, },
+    { change (_ + _) * _ = _ * _ + _ * _, rw right_distrib, },
+    repeat { set G := lift_fun R M f,
+      change G _ + G _ = G _ + G _ <|> change G _ * G _ = G _ * G _, rw h_ih, },
+    { change (algebra_map _ _ _) * _ = algebra_map _ _ _, simp },
+    { change _ * (algebra_map _ _ _) = algebra_map _ _ _, simp },
+  end,
+  map_one' := by change algebra_map _ _ _ = _; simp,
+  map_mul' := by { rintros ⟨⟩ ⟨⟩, refl },
+  map_zero' := by change algebra_map _ _ _ = _; simp,
+  map_add' := by { rintros ⟨⟩ ⟨⟩, refl },
+  commutes' := by tauto }
+
+variables {R M}
+
+@[simp]
+theorem ι_comp_lift {A : Type u₃} [semiring A] [algebra R A] (f : M →ₗ[R] A) :
+  (lift R M f).to_linear_map.comp (ι R M) = f := by {ext, refl}
+
+@[simp]
+theorem lift_unique {A : Type u₃} [semiring A] [algebra R A] (f : M →ₗ[R] A)
+  (g : tensor_algebra R M →ₐ[R] A) : g.to_linear_map.comp (ι R M) = f ↔ g = lift R M f :=
+begin
+  refine ⟨λ hyp, _, λ hyp, by simp [hyp]⟩,
+  ext,
+  rcases quot.exists_rep x with ⟨x,rfl⟩,
+  induction x,
+  { change (g.to_linear_map.comp (ι R M)) _ = _, rw hyp, refl },
+  { exact alg_hom.commutes g x },
+  repeat { change g (quot.mk _ _ + quot.mk _ _) = _ <|> change g (quot.mk _ _ * quot.mk _ _) = _,
+    rw alg_hom.map_add <|> rw alg_hom.map_mul,
+    rw [x_ih_a, x_ih_a_1], refl },
+end
+
+theorem hom_ext {A : Type u₃} [semiring A] [algebra R A] {f g : tensor_algebra R M →ₐ[R] A} :
+  f.to_linear_map.comp (ι R M) = g.to_linear_map.comp (ι R M) → f = g := λ hyp,
+  let h := g.to_linear_map.comp (ι R M) in
+  by rw [show g = lift R M h, by rw ←lift_unique, ←lift_unique, hyp]
```

Comment:
New line after the `:=`, and then no indenting required in the proof.

I'm not so keen on the `show ..., by ...` inside a `rw`, but I guess people do it. I would just `have` it first then `rw` by it.